### ISSUE TITLE
Streaming SSR

### ISSUE THREAD

**Issue by cyco130** [rakkasjs/rakkasjs#2]
React has support for streaming SSR which is being reworked for React 18. We should support it. Some potential issues:

- ~~It seems to be tied to Node's stream APIs, it may not work on worker environments (like Cloudflare's).~~ (solved in later releases)
- Some integrations (like Apollo client) may not support it, so it should be opt in.

**Comment by MarkLyck** [comment]
Any update on this?

**Comment by cyco130** [comment]
I started work on [my personal fork](https://github.com/cyco130/rakkasjs/tree/react18). It's still in exploratory phase but I can safely say that a) it works, b) it brings breaking changes in data fetching API, c) but the new API will be a lot nicer (you can fetch from any component, not just pages). Please check back soon, I'm hoping to release in a few weeks.

**Comment by cyco130** [comment]
This is working in 0.6.

**Comment by peteruithoven** [comment]
This "React 18 (server components etc.)" is mentioned as planned on: https://rakkasjs.org/guide/feature-comparison
Should that be updated if it's now released?

**Comment by IgnusG** [comment]
> Version 0.6 is currently `rakkasjs@next` and hasn't been officially released. It's still in rough alpha stage but streaming SSR with Suspense for data fetching is functional. I'm aiming for a release right after Vite 3 stable is out (should be in about a month).
> 
> Development is on the [next](https://github.com/rakkasjs/rakkasjs/tree/next) branch and there are a few [quite exciting features](https://stackblitz.com/edit/use-ssq?file=src%2Froutes%2Findex.page.tsx) you might like to check.
> 
> I closed the issue so that people don't work on a PR about it. I updated the closing comment accordingly.

There's some really cool progress on the next branch. What I'm missing from the above linked example however is the ability to imperatively call the server code to get the data from outside the render cycle of the component. 

With the next branch, if I want to fetch something again I have to set a new state in the component and then call refetch. This can lead to a lot of state juggling for complex queries. There's also no way to call these endpoints from outside components (eg. when you need more control over when and what resources are loaded forcing you to default to fetch and API routes again).

For comparison: Blitz.js has had a similar concept to the hook `useServerSideQuery` that is proposed here. Blitz does require you to specify these functions in specially named files but you can simply import them on the client and they will be magically transformed into http requests just like here. And since they are also available as simple functions they can be called from anywhere which makes them insanely ergonomic - and they can easily be wrapped into a special hook if that's needed.

**Comment by cyco130** [comment]
@IgnusG thanks for your interest!

The linked example is just a rough demonstration of the concept. This week's focus will be implementing the following plan:

There will be two imperative functions called `runServerSideQuery` and `runServerSideMutation`. `useServerSideQuery` and `useServerSideMutation` will be convenience wrappers around `useQuery`+`runServerSideQuery`  and `useMutation`+`runServerSideMutation` respectively.

`useQuery` is a minimal, Suspense-only subset of the similarly named [`react-query`](https://react-query.tanstack.com/) API and it's been already implemented. And `useMutation` will be a subset of `react-query`'s `useMutation`. I'm planning to drop my own implementations in favor of the original `react-query` ones when/if they release a streaming-compatible version.

The difference between `runServerSideQuery` and `runServerSideMutation` is that the former can be called either during SSR or on the client so it can appear on the render path and throw Suspense promises. The latter will only be available on the client and can only be called from event handlers, `useEffect` callbacks etc. Because `runServerSideQuery` can be called on the server, it requires a context parameter (which will be passed to the callback). It will also issue a `GET` request by default and it will be possible to prerender the responses for static sites.

In short, you will be able to do things imperatively, like `runServerSideMutation(() => doSomethingOnTheSever())` but I expect the `useServerSide*` hooks to be adequate for most scenarios.

I'm also hoping to provide integration examples with [telefunc](https://telefunc.com/) and [tRPC](https://trpc.io/) for more complex RPC scenarios. And finally, API routes and `fetch` will of course continue to be available for building REST or GraphQL endpoints.

**Comment by IgnusG** [comment]
> @IgnusG thanks for your interest!
> 
> The linked example is just a rough demonstration of the concept. This week's focus will be implementing the following plan:
> 
> There will be two imperative functions called `runServerSideQuery` and `runServerSideMutation`. `useServerSideQuery` and `useServerSideMutation` will be convenience wrappers around `useQuery`+`runServerSideQuery` and `useMutation`+`runServerSideMutation` respectively.
> 
> `useQuery` is a minimal, Suspense-only subset of the similarly named [`react-query`](https://react-query.tanstack.com/) API and it's been already implemented. And `useMutation` will be a subset of `react-query`'s `useMutation`. I'm planning to drop my own implementations in favor of the original `react-query` ones when/if they release a streaming-compatible version.
> 
> The difference between `runServerSideQuery` and `runServerSideMutation` is that the former can be called either during SSR or on the client so it can appear on the render path and throw Suspense promises. The latter will only be available on the client and can only be called from event handlers, `useEffect` callbacks etc. Because `runServerSideQuery` can be called on the server, it requires a context parameter (which will be passed to the callback). It will also issue a `GET` request by default and it will be possible to prerender the responses for static sites.
> 
> In short, you will be able to do things imperatively, like `runServerSideMutation(() => doSomethingOnTheSever())` but I expect the `useServerSide*` hooks to be adequate for most scenarios.
> 
> I'm also hoping to provide integration examples with [telefunc](https://telefunc.com/) and [tRPC](https://trpc.io/) for more complex RPC scenarios. And finally, API routes and `fetch` will of course continue to be available for building REST or GraphQL endpoints.

Ah perfect! I'm not a big fan of react-query which is why I was mentioning the imperative approach. `runServerSideQuery` and `runServerSideMutation` are pretty much the exact thing I was hoping for (similar to what Blitz offers with its query and mutation handlers - but of course they are built on top of NextJS - whereas I prefer Vite).

I'm curious about the `tRPC` integration too! I was quite happy with it last time I used it.

**Comment by cyco130** [comment]
> I'm curious about the `tRPC` integration too! I was quite happy with it last time I used it.

It should be usable even today. I know [someone was working on it](https://github.com/sachinraja/trpc-rakkas) at some point but I've never tried it myself.

**Comment by cyco130** [comment]
> I'm not a big fan of react-query

I believe a framework like Rakkas has to come with something that works out of the box. To my knowledge Relay is the only library that has semi-official streaming SSR with Suspense support. So I had to come up with something, `react-query` is popular and easy to use so I based my interim solution on that. As more and more libraries start to support streaming with Suspense, hopefully we can make more and more data fetching methods available.

Of course you can use pretty much anything today if you don't need SSR though.

**Comment by IgnusG** [comment]
> > I'm not a big fan of react-query
> 
> I believe a framework like Rakkas has to come with something that works out of the box. To my knowledge Relay is the only library that has semi-official streaming SSR with Suspense support. So I had to come up with something, `react-query` is popular and easy to use so I based my interim solution on that. As more and more libraries start to support streaming with Suspense, hopefully we can make more and more data fetching methods available.
> 
> Of course you can use pretty much anything today if you don't need SSR though.

Yes that's understandable. One of the goals of rakkas is after all to provide an out-of-the-box opinionated experience. 
Can't wait for future updates on `rakkasjs@next`!

**Comment by cyco130** [comment]
@IgnusG version `0.6.0-next.6` supports `import * as rakkas from "rakkasjs"`. Also implemented `runServerSideMutation` and `useMutation`. `runServerSideMutation` is can be used for client-only data fetching.
